kinode/packages/contacts/contacts/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+use crate::kinode::process::contacts;
+use kinode_process_lib::{
+    await_message, call_init, eth, get_blob, get_typed_state, homepage, http, kimap, kiprintln,
+    set_state, Address, Capability, LazyLoadBlob, Message, NodeId, Response,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::str::FromStr;
+
+const ICON: &str = include_str!("icon");
+
+#[cfg(not(feature = "simulation-mode"))]
+const CHAIN_ID: u64 = kimap::KIMAP_CHAIN_ID;
+#[cfg(feature = "simulation-mode")]
+const CHAIN_ID: u64 = 31337; // local
+
+const CHAIN_TIMEOUT: u64 = 60; // 60s
+
+#[cfg(not(feature = "simulation-mode"))]
+const KIMAP_ADDRESS: &'static str = kimap::KIMAP_ADDRESS; // optimism
+#[cfg(feature = "simulation-mode")]
+const KIMAP_ADDRESS: &str = "0xEce71a05B36CA55B895427cD9a440eEF7Cf3669D";
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Contact(HashMap<String, serde_json::Value>);
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Contacts(HashMap<NodeId, Contact>);
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ContactsState {
+    our: Address,
+    contacts: Contacts,
+}
+
+impl ContactsState {
+    fn new(our: Address) -> Self {
+        get_typed_state(|bytes| serde_json::from_slice(bytes)).unwrap_or(Self {
+            our,
+            contacts: Contacts(HashMap::new()),
+        })
+    }
+
+    fn save(&self) {
+        set_state(&serde_json::to_vec(&self).expect("Failed to serialize contacts state!"));
+    }
+
+    fn contacts(&self) -> &Contacts {
+        &self.contacts
+    }
+
+    fn get_contact(&self, node: NodeId) -> Option<&Contact> {
+        self.contacts.0.get(&node)
+    }
+
+    fn add_contact(&mut self, node: NodeId) {
+        self.contacts.0.insert(node, Contact(HashMap::new()));
+        self.save();
+    }
+
+    fn remove_contact(&mut self, node: NodeId) {
+        self.contacts.0.remove(&node);
+        self.save();
+    }
+
+    fn add_field(&mut self, node: NodeId, field: String, value: serde_json::Value) {
+        self.contacts
+            .0
+            .entry(node)
+            .or_insert_with(|| Contact(HashMap::new()))
+            .0
+            .insert(field, value);
+        self.save();
+    }
+
+    fn remove_field(&mut self, node: NodeId, field: String) {
+        if let Some(contact) = self.contacts.0.get_mut(&node) {
+            contact.0.remove(&field);
+        }
+        self.save();
+    }
+
+    fn ws_update(&self, http_server: &mut http::server::HttpServer) {
+        http_server.ws_push_all_channels(
+            "/",
+            http::server::WsMessageType::Text,
+            LazyLoadBlob::new(
+                Some("application/json"),
+                serde_json::to_vec(self.contacts()).unwrap(),
+            ),
+        );
+    }
+}
+
+wit_bindgen::generate!({
+    path: "target/wit",
+    world: "contacts-sys-v0",
+    generate_unused_types: true,
+    additional_derives: [PartialEq, serde::Deserialize, serde::Serialize],
```

Comment:
Consider using `process_macros::SerdeJsonInto` which will reduce `serde_json` usage below. I'd be happy to do a pass through and convert over to using if you'd like

Reply:
good idea, i'll do it to get in the habit